### PR TITLE
[w5.4][T14-A1]Xu Jun

### PR DESCRIPTION
**Pull request by willxujun**

### CODE REVIEW

**Comment by willxujun** [comment]
This big PR contains changes I made over the last two weeks but for which I failed to submit separate PRs, including:
[w4.3d] splitting Address class
[w4.4] RuntimeException handling
[w5.2] One test for Util
[w5.4] Add undo command (pending documentation)

One LO still in progress:
[w5.3] Inheritance

File: src/seedu/addressbook/commands/UndoCommand.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package seedu.addressbook.commands;
+
+import seedu.addressbook.data.person.Person;
+import seedu.addressbook.data.stack.CommandPair;
+import static seedu.addressbook.ui.TextUi.DISPLAYED_INDEX_OFFSET;
+
+public class UndoCommand extends Command {
+    public static final String COMMAND_WORD = "undo";
+    private Person person;
+    private int isUndoingAdd;
```
Comment (Reginleiff) [inline src/seedu/addressbook/commands/UndoCommand.java]: Use a boolean instead for better clarity on line 38?

**Comment by Reginleiff** [comment]
I am thinking perhaps how we can better prepare the undo function to undo sorts? Perhaps it may be a better idea to store the state of the environment before each operation and just revert it back to that state upon the undo call?

File: src/seedu/addressbook/data/stack/CommandPair.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package seedu.addressbook.data.stack;
+
+import seedu.addressbook.data.person.Person;
+
+public class CommandPair {
```
Comment (K1ang) [inline src/seedu/addressbook/data/stack/CommandPair.java]: Missing header comment. Remember to include header comment for classes and nontrivial methods to improve the clarity of the code.

File: src/seedu/addressbook/commands/UndoCommand.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            AddCommand add = new AddCommand(person);
+            result = reExecuteCommand(add);
+        } else {
+            int index = addressBook.findIndexOf(person);
+            DeleteCommand delete = new DeleteCommand(index);
+            result = reExecuteCommand(delete);
+        }
+        return result;
+    }
+    */
```
Comment (K1ang) [inline src/seedu/addressbook/commands/UndoCommand.java]: Don't leave commented out code. Just go ahead and delete unused code. Now that we are using a revision control system, we can recover any deleted code later.

**Comment by K1ang** [comment]
Some comments added. Please close the PR after reading comments.

File: src/seedu/addressbook/data/stack/CommandPair.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (willxujun) [reply src/seedu/addressbook/data/stack/CommandPair.java]: This is for undo stack, containing a Person and a Int flag indicating whether this Person is previously deleted or added. Whenever undo is cast, pop a CommandPair from stack, fetch the Person, read the flag so as to generate the reverse operation

File: src/seedu/addressbook/data/stack/CommandPair.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    private Person person;
+    private int isAdd;
+
+    public CommandPair(Person person, int isAdd) {
+        this.person = person;
+        this.isAdd = isAdd;
```
[the remark this answers is not in the document]
Comment (willxujun) [reply src/seedu/addressbook/data/stack/CommandPair.java]: Yes it is intended to be a boolean. I somehow used int. Nevertheless it still functions like a boolean flag.

**Pull request closed without merging**